### ISSUE TITLE
Use as a library

### ISSUE THREAD

**Issue by miromannino** [BayesianLogic/blog#342]
Let's say I have a server, written in Java, and I need to execute the same model with some specific observation.

I would like to use BLOG as a library instead of using it as an external command line application.
How can I do that? I have seen something that could be used in the test folder, for example using `simpleSetupFromFiles`; but I cannot find any official/compact solution.

Thank you

**Comment by jxwuyi** [comment]
@miromannino  Hi, you can always execute BLOG by (1) running command line inside a process (2) call the Java class if you are using Java or Scala. 
The _simpleSetupFromFiles_ option does essentially the same thing as the command line: it takes in the model/query/obs written in a text file. 
You can always write a model in a fixed text file and generate different query files and obs files.
I guess the subtle thing is data/query input and output.
BLOG can output to jason format. 
So what kind of input/output would you like?
I mean the most naive solution will be sth like: you write a script which takes in the data (in some format) and produces a query/obs file and then call the BLOG class. Anyway, I think you may want nicer solution.
We are working on data input/output stuff recently. So we do appreciate your feedback. Can you provide mode details on your use case?

**Comment by miromannino** [comment]
Yes, the problem is that I would like to avoid launching a process to execute it. I would like to load the model, and then execute that each time with many different observations.

The Json option, seems only an option to let it to write into a file. I would avoid to do that, at least, if I need to open another process to launch BLOG, I should get the data redirecting the stdout of that process.

So how can I call it following the second strategy you were talking about?

Thank you

**Comment by jxwuyi** [comment]
Unfortunately, now the BLOG system does not support streaming/incremental data processing. You need to store the model in a file, data in a file, obs in a file; then load the java class; process all the fies to process a internal data structure; and perform inference as a whole over all the files.
Only compiling the model and support incremental/streaming/new data will be a good direction for us. We will start to investigate this option in the near future.

Well, for the output, the output from screen will be the statistics of all the produced samples. If you need all the produced samples, there is currently now way to load that. The samples are stored in a internal histogram data structure. The jason file will contain all the produced samples.

So can you tell me what form you would like the output data in? I can probably add a new output option for you.
